### ISSUE TITLE
Error: Unable to open netmap:eth0: Invalid argument in AWS

### ISSUE THREAD

**Issue by gsaneryeeb** [luigirizzo/netmap#566]
Hello, I installed netmap in AWS(Ubuntu 16.04) .
When I run `$ pke-gen -i eth0` , I got these error message:
```
065.096883 main [2855] interface is eth0
065.097475 main [2977] using default burst size: 512
065.097579 main [2985] running on 1 cpus (have 1)
065.097753 extract_ip_range [467] range is 10.0.0.1:1234 to 10.0.0.1:1234
065.097828 extract_ip_range [467] range is 10.1.0.1:1234 to 10.1.0.1:1234
065.097897 nm_open [855] overriding ARG1 0
065.097932 nm_open [859] overriding ARG2 0
065.097964 nm_open [863] overriding ARG3 0
065.097996 nm_open [867] overriding RING_CFG
065.098027 nm_open [876] overriding ifname eth0 ringid 0x0 flags 0x8001
065.128843 nm_open [947] NIOCREGIF failed: Invalid argument netmap:eth0
065.128903 main [3078] Unable to open netmap:eth0: Invalid argument
065.128937 main [3145] aborting
``` 
Is there any special settings when I use the virtual server in AWS, GCP, etc
Thanks.

**Comment by vmaffione** [comment]
Hi,
  There are no special settings for AWS, GCP, etc.

Does `dmesg` report any netmap-related error or info?
What is the output of `ethtool -i eth0`?
Which commit id are you using (e.g. the SHA sum)?

**Comment by vmaffione** [comment]
Hi,
  The dmesg clearly shows what the problem is.
You are using the emulated netmap adapter on top of a `vif` interface (Xen netfront paravirtualized NIC).
However, the MTU of your network interface is set to 9000, which does not play well with the netmap buffer size (2048 bytes by default). If emulated adapter supported multi-fragment operation (`NS_MOREFRAG`), this configuration would be ok. However, since `NS_MOREFRAG` is not supported yet on the emulated adapter, you need the netmap buffer size to be `>=` of the MTU.
In other words, the simplest solution is to lower your MTU:
````
 # ip link set dev eth0 mtu 2000
````
Or you could also modify the netmap buffer size to satistfy the inequality
````
 # echo ${NUM_BYTES} > /sys/module/netmap/parameters/buf_size
````

(The commit id does not matter in this case. You can get it with `git log` in the netmap repo, and looking at the hex number on top, e.g. `commit fce88df1b49c0dfce4c6d4211f1d4ec62aa320b0`.)

**Comment by vmaffione** [comment]
The `drop mbuf` messages are caused by the NIC offloadings not being disabled. Look at https://github.com/luigirizzo/netmap/blob/master/LINUX/README#L243-L255 to see how to disable them.

Regarding the freeze, if `eth0` is the only interface of your VM, that's expected.
When you open `eth0` in netmap mode (e.g. by running `pkt-gen` or other netmap applications), `eth0` gets "disconnected" from the network stack, so that it can be used by netmap. As a result, network traffic does not flow between `eth0` and your SSH server, which means that your ssh session will freeze.

**Comment by vmaffione** [comment]
You can run the unit tests and integration tests if you wish:
````
$ sudo make unitest
$ sudo make intest
````

What do you mean by "set only open port 80 and 22"?

**Comment by gsaneryeeb** [comment]
It means that I create rules for ipfw to allow only ports 80/tcp and 22/tcp.
Like this:
```
# The rules
$cmd 01000 allow tcp from any to any 80,22
$cmd 01500 allow tcp from any 80,22 to any
```
My solution is to install netmap and [dummynet](https://github.com/luigirizzo/dummynet) and implement it by setting the rules for ipfw. But I can't build in Ubuntu 16.04 with kernel 4.4.0
The error is:

```
loki@ip-172-31-36-20:~/dummynet$ make KERNELPATH=/usr/src/linux-headers-4.4.0-1061-aws
make[1]: Entering directory '/home/loki/dummynet/kipfw-mod'
Makefile:76: ------ arch Linux goals include_e -----------
Makefile:202: ------------- linux version 40483 (hex) ------------
Makefile:264: --- build 2.6 and newer target kipfw ----
Makefile:289:
make[1]: Leaving directory '/home/loki/dummynet/kipfw-mod'
make[1]: Entering directory '/home/loki/dummynet/kipfw-mod'
Makefile:76: ------ arch Linux goals kipfw -----------
Makefile:202: ------------- linux version 40483 (hex) ------------
Makefile:264: --- build 2.6 and newer target kipfw ----
Makefile:289:
echo "xxxxxxxxxxxxx make -C /usr/src/linux-headers-4.4.0-1061-aws V= M=`pwd` modules"
xxxxxxxxxxxxx make -C /usr/src/linux-headers-4.4.0-1061-aws V= M=/home/loki/dummynet/kipfw-mod modules
make -C /usr/src/linux-headers-4.4.0-1061-aws V= M=`pwd` modules
make[2]: Entering directory '/usr/src/linux-headers-4.4.0-1061-aws'
/home/loki/dummynet/kipfw-mod/Makefile:76: ------ arch Linux goals  -----------
/home/loki/dummynet/kipfw-mod/Makefile:202: ------------- linux version 40483 (hex) ------------
/home/loki/dummynet/kipfw-mod/Makefile:264: --- build 2.6 and newer target kipfw ----
/home/loki/dummynet/kipfw-mod/Makefile:289:
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw2.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw_pfil.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw_sockopt.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw_dynamic.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw_table.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_fw_log.o
  CC [M]  /home/loki/dummynet/kipfw-mod/radix.o
  CC [M]  /home/loki/dummynet/kipfw-mod/in_cksum.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_dummynet.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_dn_io.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ip_dn_glue.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_heap.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_sched_fifo.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_sched_wf2q.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_sched_rr.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_sched_qfq.o
  CC [M]  /home/loki/dummynet/kipfw-mod/dn_sched_prio.o
  CC [M]  /home/loki/dummynet/kipfw-mod/ipfw2_mod.o
In file included from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:100:0:
include/net/inet_hashtables.h: In function 'sk_daddr_set':
include/net/inet_hashtables.h:354:2: error: implicit declaration of function 'ipv6_addr_set_v4mapped' [-Werror=implicit-function-declaration]
  ipv6_addr_set_v4mapped(addr, &sk->sk_v6_daddr);
  ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c: In function 'ip_output':
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:362:15: error: passing argument 1 of 'dst_output' from incompatible pointer type [-Werror=incompatible-pointer-types]
    dst_output(skb);
               ^
In file included from include/net/sock.h:67:0,
                 from include/linux/tcp.h:22,
                 from include/linux/ipv6.h:72,
                 from include/net/netfilter/nf_queue.h:5,
                 from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:
include/net/dst.h:493:19: note: expected 'struct net *' but argument is of type 'struct sk_buff *'
 static inline int dst_output(struct net *net, struct sock *sk, struct sk_buff *skb)
                   ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:362:4: error: too few arguments to function 'dst_output'
    dst_output(skb);
    ^
In file included from include/net/sock.h:67:0,
                 from include/linux/tcp.h:22,
                 from include/linux/ipv6.h:72,
                 from include/net/netfilter/nf_queue.h:5,
                 from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:
include/net/dst.h:493:19: note: declared here
 static inline int dst_output(struct net *net, struct sock *sk, struct sk_buff *skb)
                   ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c: In function 'ipfw2_queue_handler':
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:559:26: error: 'struct nf_queue_entry' has no member named 'indev'
  m->m_pkthdr.rcvif = info->indev;
                          ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:568:10: error: 'struct nf_queue_entry' has no member named 'hook'
  if (info->hook == IPFW_HOOK_IN) {
          ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:569:39: error: 'struct nf_queue_entry' has no member named 'indev'
   ret = ipfw_check_hook(NULL, &m, info->indev, PFIL_IN, NULL);
                                       ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:571:39: error: 'struct nf_queue_entry' has no member named 'outdev'
   ret = ipfw_check_hook(NULL, &m, info->outdev, PFIL_OUT, NULL);
                                       ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c: At top level:
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:827:35: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
                 .hook           = call_ipfw,
                                   ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:827:35: note: (near initialization for 'ipfw_ops[0].hook')
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:831:17: error: unknown field 'owner' specified in initializer
                 SET_MOD_OWNER
                 ^
In file included from include/linux/linkage.h:6:0,
                 from include/linux/kernel.h:6,
                 from include/linux/list.h:8,
                 from include/linux/msg.h:4,
                 from /home/loki/dummynet/kipfw-mod/../glue.h:101,
                 from <command-line>:0:
include/linux/export.h:34:21: error: excess elements in struct initializer [-Werror]
 #define THIS_MODULE (&__this_module)
                     ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:821:32: note: in expansion of macro 'THIS_MODULE'
 #define SET_MOD_OWNER .owner = THIS_MODULE,
                                ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:831:17: note: in expansion of macro 'SET_MOD_OWNER'
                 SET_MOD_OWNER
                 ^
include/linux/export.h:34:21: note: (near initialization for 'ipfw_ops[0]')
 #define THIS_MODULE (&__this_module)
                     ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:821:32: note: in expansion of macro 'THIS_MODULE'
 #define SET_MOD_OWNER .owner = THIS_MODULE,
                                ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:831:17: note: in expansion of macro 'SET_MOD_OWNER'
                 SET_MOD_OWNER
                 ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:834:35: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
                 .hook           = call_ipfw,
                                   ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:834:35: note: (near initialization for 'ipfw_ops[1].hook')
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:838:3: error: unknown field 'owner' specified in initializer
   SET_MOD_OWNER
   ^
In file included from include/linux/linkage.h:6:0,
                 from include/linux/kernel.h:6,
                 from include/linux/list.h:8,
                 from include/linux/msg.h:4,
                 from /home/loki/dummynet/kipfw-mod/../glue.h:101,
                 from <command-line>:0:
include/linux/export.h:34:21: error: excess elements in struct initializer [-Werror]
 #define THIS_MODULE (&__this_module)
                     ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:821:32: note: in expansion of macro 'THIS_MODULE'
 #define SET_MOD_OWNER .owner = THIS_MODULE,
                                ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:838:3: note: in expansion of macro 'SET_MOD_OWNER'
   SET_MOD_OWNER
   ^
include/linux/export.h:34:21: note: (near initialization for 'ipfw_ops[1]')
 #define THIS_MODULE (&__this_module)
                     ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:821:32: note: in expansion of macro 'THIS_MODULE'
 #define SET_MOD_OWNER .owner = THIS_MODULE,
                                ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:838:3: note: in expansion of macro 'SET_MOD_OWNER'
   SET_MOD_OWNER
   ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c: In function 'ipfw_module_init':
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:780:28: error: passing argument 1 of 'nf_register_queue_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
 #define REG_QH_ARG(pf, fn) &(fn ## _desc)
                            ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:904:32: note: in expansion of macro 'REG_QH_ARG'
      nf_register_queue_handler(REG_QH_ARG(PF_INET, ipfw2_queue_handler) );
                                ^
In file included from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:0:
include/net/netfilter/nf_queue.h:31:6: note: expected 'struct net *' but argument is of type 'struct nf_queue_handler *'
 void nf_register_queue_handler(struct net *net, const struct nf_queue_handler *qh);
      ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:904:6: error: too few arguments to function 'nf_register_queue_handler'
      nf_register_queue_handler(REG_QH_ARG(PF_INET, ipfw2_queue_handler) );
      ^
In file included from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:0:
include/net/netfilter/nf_queue.h:31:6: note: declared here
 void nf_register_queue_handler(struct net *net, const struct nf_queue_handler *qh);
      ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:918:2: error: too few arguments to function 'nf_unregister_queue_handler'
  nf_unregister_queue_handler(UNREG_QH_ARG(PF_INET, ipfw2_queue_handler) );
  ^
In file included from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:0:
include/net/netfilter/nf_queue.h:32:6: note: declared here
 void nf_unregister_queue_handler(struct net *net);
      ^
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c: In function 'ipfw_module_exit':
/home/loki/dummynet/kipfw-mod/ipfw2_mod.c:942:2: error: too few arguments to function 'nf_unregister_queue_handler'
  nf_unregister_queue_handler(UNREG_QH_ARG(PF_INET, ipfw2_queue_handler) );
  ^
In file included from /home/loki/dummynet/kipfw-mod/ipfw2_mod.c:71:0:
include/net/netfilter/nf_queue.h:32:6: note: declared here
 void nf_unregister_queue_handler(struct net *net);
      ^
cc1: all warnings being treated as errors
scripts/Makefile.build:277: recipe for target '/home/loki/dummynet/kipfw-mod/ipfw2_mod.o' failed
make[3]: *** [/home/loki/dummynet/kipfw-mod/ipfw2_mod.o] Error 1
Makefile:1437: recipe for target '_module_/home/loki/dummynet/kipfw-mod' failed
make[2]: *** [_module_/home/loki/dummynet/kipfw-mod] Error 2
make[2]: Leaving directory '/usr/src/linux-headers-4.4.0-1061-aws'
Makefile:310: recipe for target 'kipfw' failed
make[1]: *** [kipfw] Error 2
make[1]: Leaving directory '/home/loki/dummynet/kipfw-mod'
Makefile:110: recipe for target 'kipfw' failed
make: *** [kipfw] Error 2
```

**Comment by vmaffione** [comment]
But netmap and dummynet are unrelated.
Btw you get those errors because nobody is actively maintaining dummynet (since 2015), so it does not build on newer kernels.

The good news is that you can implement such a simple filter very efficiently with a simple netmap application, starting from the `bridge` example application, and adding some logic to filter out the unwanted packets by looking inside the packet headers.
I would recommend looking at our tutorial to learn how to write simple netmap applications https://github.com/netmap-unipi/netmap-tutorial

**Comment by vmaffione** [comment]
I think so. You can probably use something like
````
./kipfw netmap:eth0 netmap:eth0^
````
because you want to apply the filter between the eth0 NIC and the network stack.

**Comment by gsaneryeeb** [comment]
Thank you for you advice.

1.  For this command `kipfw`, where can we look for more detail information about how to use it?
2. Because I used an AWS server and used `SSH` (port 22 ) to log in the server, can I test the commands of netmap-ipfw on eth0 ?  I run the command `$ pkt-gen -i eth0` on previous test, I lost my connection with my server for a while.
3. I run the example code below. It looks like work. But I don't know where the ipfw rule files save, is it a dynamic rule？Will it be lost, when I reboot the server.

```
# connect the firewall to two vale switches
	./kipfw valeA:f valeB:f &

	# configure ipfw/dummynet
	ipfw/ipfw show	# or other

	# start the sink
	pkt-gen -i valeB:d -f rx

	# start an infinite source
	pkt-gen -i valeA:s -f tx

	# plain again with the firewall and enjoy
	ipfw/ipfw show  # or other
```

**Comment by vmaffione** [comment]
The `kipfw` program is experimental, so there is no documentation in addition to the `README`.
I explained above the reason why `pkt-gen -i eth0` cause the termination/freeze of your SSH connection: nothing is passing to/from the network stack.
If you run `./kipfw  netmap:eth0 netmap:eth0^`, you are creating a firewall between the eth0 NIC rings (`netmap:eth0') and the eth0 interface visible from the network stack (`netmap:eth0^').
The default ruleset is "pass any to any", as you can see running `ipfw/ipfw show`.
You can change the ruleset using the `ipfw/ipfw` command: you can find the documentation in the FreeBSD man page `ipfw(8)` (https://www.freebsd.org/cgi/man.cgi?ipfw(8)).

(Closing this as there is no issue, but more questions are welcome).

**Comment by vmaffione** [comment]
I don't think the ruleset is persistent; as I said, `kipfw` is experimental. You can always arrange a startup script that builds the rules you need.

Another option you have is to modify this program https://github.com/netmap-unipi/netmap-tutorial/blob/master/solutions/forward.c#L53-L81 (part of the netmap tutorial).
If you modify the `pkt_select` function to match TCP (instead of UDP) and ports 20 or 80 (instead of the port contained in `udp_port`), you will get the functionality you are looking for (`./forward -i netmap:eth0 -i netmap:eth0^`).

**Comment by gsaneryeeb** [comment]
Very Helpful !

**Comment by gsaneryeeb** [comment]
I try to run `./kipfw netmap:eth0 netmap:eth0^` on my server. I can see the default ruleset is `65535 0 0 allow ip from any to any`.  And I also added ruleset `allow tcp from any to any`.  But I still can't use SSH client to connect this server. The error is `ssh_exchange_identification: read: Connection reset by peer`

**Comment by vmaffione** [comment]
Maybe rules for letting ARP are missing? IOW, can you ping the VM from outside?

**Comment by vmaffione** [comment]
Sorry, I don't understand how you did your tests.
What I meant is to check if ping and or SSH (both from the outside) is working in the three following cases:
A: no netmap application is running in the VM
B: while ./bridge is running as above
C: while ./kipfw is running as above.

**Comment by gsaneryeeb** [comment]
But after case C, I add rules like this 
```
$ sudo ipfw/ipfw add allow icmp from me to any icmptypes 8 out
$ sudo ipfw/ipfw add allow icmp from any to me icmptypes 0 in
$ sudo ipfw/ipfw show
00100  0    0 allow icmp from me to any icmptypes 8 out
00200  0    0 allow icmp from any to me icmptypes 0 in
65535 38 3480 allow ip from any to any
```
The rules are invalid. I can still ping from the outside.

**Comment by gsaneryeeb** [comment]
Hi, vmaffione.
Yes, The aims of these two rules is that I can ping from VM to outside and I can't ping from outside to VM.

**Comment by vmaffione** [comment]
If I were you I would ask on the freebsd-net mailing list, to check if the rules are correct.

**Comment by gsaneryeeb** [comment]
Sure, I added these rules on FreeBSD 11.2, It's workable.

**Comment by gsaneryeeb** [comment]
As you said before, the kipfw program is experimental. So maybe it can't work in this situation.

**Comment by vmaffione** [comment]
Ok, also consider that the ipfw code was taken on `2012-08-03` (as you can read in the netmap-ipfw README), so that's a lot of time ago. It may be significantly different from FreeBSD 11.2.

Probably you can debug it a little bit to understand what is going on.
Or make your own application based on the `forward` program (tutorial).
